Key Vault mgmt commands

derekbekoe opened the pull request:
**Key Vault create**
```
$ az keyvault create --help

Command
    az keyvault create

Arguments
    --location -l          [Required]: Location.
    --name -n              [Required]: The primary resource name.
    --resource-group -g    [Required]: Name of resource group.
    --enabled-for-deployment         : Property to specify whether Azure Virtual Machines are
                                       permitted to retrieve certificates stored as secrets from the
                                       key vault.
    --enabled-for-disk-encryption    : Property to specify whether Azure Disk Encryption is
                                       permitted to retrieve secrets from the vault and unwrap keys.
    --enabled-for-template-deployment: Property to specify whether Azure Resource Manager is
                                       permitted to retrieve secrets from the key vault.
    --no-self-perms                  : If specified, don't add permissions for the current user in
                                       the new vault.
    --sku                            : SKU details.  Allowed values: premium, standard.  Default:
                                       standard.
    --tags
```

**Key Vault list**
```
$ az keyvault list --help

Command
    az keyvault list

Arguments
    --resource-group -g: Name of resource group.
```

**Key Vault show**
```
$ az keyvault show --help

Command
    az keyvault show

Arguments
    --ids              : ID of resource.
    --name -n          : The primary resource name.
    --resource-group -g: Name of resource group.
```

**Key Vault delete**
```
$ az keyvault delete --help

Command
    az keyvault delete

Arguments
    --ids              : ID of resource.
    --name -n          : The primary resource name.
    --resource-group -g: Name of resource group.
```
**Key Vault update**
```
$ az keyvault update --help

Command
    az keyvault update

Arguments
    --add              : Add an object to a list of objects by specifying a path and key value
                         pairs.  Example: --add property.list key=<value>.
    --ids              : ID of resource.
    --name -n          : The primary resource name.
    --remove           : Remove a property or an element from a list.  Example: --remove
                         property.list <index>.
    --resource-group -g: Name of resource group.
    --set              : Update an object by specifying a property path and value to set.  Example:
                         --set property1.property2=value.
```

**Key Vault set policy**
```
$ az keyvault set-policy  --help

Command
    az keyvault set-policy

Arguments
    --ids              : ID of resource.
    --name -n          : The primary resource name.
    --object-id        : A GUID that identifies the principal that will receive permissions.
    --perms-to-keys    : Permissions to keys.
    --perms-to-secrets : Permissions to secrets.
    --resource-group -g: Name of resource group.
    --spn              : Name of a service principal that will receive permissions.
    --upn              : Name of a user principal that will receive permissions.
```

**Key Vault delete policy**
```
$ az keyvault delete-policy  --help

Command
    az keyvault delete-policy

Arguments
    --ids              : ID of resource.
    --name -n          : The primary resource name.
    --object-id        : A GUID that identifies the principal that will receive permissions.
    --resource-group -g: Name of resource group.
    --spn              : Name of a service principal that will receive permissions.
    --upn              : Name of a user principal that will receive permissions.
```

tjprescott commented on src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/__init__.py:
```diff
@@ -0,0 +1,11 @@
+
+import azure.cli.command_modules.keyvault.custom
+import azure.cli.command_modules.keyvault._params
+import azure.cli.command_modules.keyvault.generated
```
>mport azure.cli.command_modules.keyvault [](start = 1, length = 40)

You should also import _help here since you added the file. This ensures that if anything is added, it will be applied.

tjprescott commented on src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/_validators.py:
```diff
@@ -0,0 +1,21 @@
+    # https://docs.python.org/2/library/argparse.html#mutual-exclusion
+    if sum(1 for p in [ns.object_id, ns.spn, ns.upn] if p) != 1:
+        raise argparse.ArgumentError(
+            None, 'One of the arguments --object-id --spn --upn is required')
```
>One of the arguments --object-id --spn --upn is required [](start = 19, length = 56)

This error message is not quite clear for the case where more than one options are specified. I would add a second error message for if sum > 2 that says "Please specify only -a, -b, or -c."

tjprescott commented on src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/_params.py:
```diff
@@ -0,0 +1,29 @@
+register_cli_argument('keyvault', 'vault_name', arg_type=name_type, completer=get_resource_name_completion_list('Microsoft.KeyVault/vaults'), id_part='name')
+register_cli_argument('keyvault create', 'vault_name', completer=None)
+
+register_cli_argument('keyvault create', 'sku_family', completer=get_enum_type_completion_list(SkuFamily), choices=[e.value for e in SkuFamily])
```
>completer=get_enum_type_completion_list(SkuFamily) [](start = 55, length = 50)

These are not necessary. If you have a choice list, it doubles as the completer. The purpose of "get_enum_type_completion_list" is for those obnoxious "model_as_string=True" swagger parameters where we can't use the enum as a choice list and thus have to expose it as a completer.

tjprescott commented on src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/_params.py:
```diff
@@ -0,0 +1,29 @@
+
+register_cli_argument('keyvault create', 'sku_family', completer=get_enum_type_completion_list(SkuFamily), choices=[e.value for e in SkuFamily])
+register_cli_argument('keyvault create', 'sku_name', completer=get_enum_type_completion_list(SkuName), choices=[e.value for e in SkuName])
+register_cli_argument('keyvault create', 'sku_name', completer=get_enum_type_completion_list(SkuName), choices=[e.value for e in SkuName])
```
Duplicate line.

tjprescott commented:
Looking at the help output, there are some params that need help text. For example, I don't what on earth 'no-self-perms' is supposed to be. From the code it looks like it relates to the KeyVault permissions.

tjprescott commented:
:clock1:

derekbekoe commented:
I've added more help text and updated the description for the PR to indicate the new help.

---
In reply to: [239268298](https://github.com/Azure/azure-cli/pull/685#issuecomment-239268298) [](ancestors = 239268298)

derekbekoe commented:
@tjprescott Feedback addressed

yugangw-msft commented on src/command_modules/azure-cli-keyvault/azure/cli/command_modules/keyvault/custom.py:
```diff
@@ -0,0 +1,182 @@
+                                        'backup',
+                                        'restore'],
+                                  secrets=['all'])
+        object_id = _get_current_user_object_id(graph_client)
```
Just a question, if you currently log in as a service principal, does this call work?

derekbekoe replied:
I now catch the error that can occur when attempting to get the current user.
Now, it works with service principal and live id.

yugangw-msft commented:
LGTM. feel free to merge after have service principle and live id scenario tested.
